[14.0][MIG] base_multi_company

rousseldenis opened the pull request:
- [x] Use odoo-test-helper to test Abstract class
- [x] Added an inverse on company_id to avoid warnings in logs
- [x] Added a depends_context on company 
- [x] Cherry-picked #313 
- [x] Simplified inverse_company_id to update once

lasley committed:
[ADD] base_multi_company: Create new module
* Create new module to provide base multi company logic and mixin
* Add deactivation by company mixin
* Add company_id/ids handling
* Add break after company is found
Squashed commits:
[854cc36] Increase test coverage
[770bd71] Revert hook view create back to model init
[40e803e] Fix apples and oranges
[7a4dfb4] Use registry correctly
[6e9f170] Switch company_id to computed & move company aliased view creation to post init hook
[faa4fc9] Remove active functionality
[fecfb59] Add explicit tests for active and inactive searches

lmignon committed:
10.0 base multi company lmi (#1)

* Revert "Revert hook view create back to model init"

This reverts commit 770bd71a919e5801702f0fd9d398a37c08472a70.

* [FIX] base_multi_company: Always create the view into a pre_init_hook to avoid error in log

lasley committed:
PR fixes (+1 squashed commit)
Squashed commits:
[4c17d04] auto_join company_ids

lmignon committed:
[FIX] base_multi_company: Allow child_of/parent_of operator into search domain on company_id/company_ids (+1 squashed commit)
Squashed commits:
[fe161fe] [ADD] setup.py

Commit:
- Remove depends as company_id is not stored anymore
- Test if a company is set in inverse method (+1 squashed commit)
Squashed commits:
[d670f30] [FIX] fix init hooks as company_id is not stored anymore

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
[IMP] base_multi_company: Improve ReadMe
* Add implementation instructions to ReadMe

pedrobaeza committed:
[MIG] *_multi_company: OpenUpgrade migration scripts

acsonefho committed:
Edit unit test + fix bug

[FIX] Fix issue based on the computation of company_id for multi_company_abstract models

[FIX] Fix issue based on the computation of company_id for multi_company_abstract models

[FIX] Fix issue based on the computation of company_id for multi_company_abstract models

[FIX] Bad branch version pushed before

[FIX] Bad branch version pushed before

[FIX] manifest

Fix flake

Flake8 on OCA tests

SimoRubi committed:
[FIX] Insert in company_ids relation's table only ids of records that are bound to some company

oca-travis committed:
[UPD] Update base_multi_company.pot

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 88.9% (8 of 9 strings)

Translation: multi-company-10.0/multi-company-10.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-10-0/multi-company-10-0-base_multi_company/pt/

rodrigets committed:
[MIG] base_multi_company: Migration to 11.0

oca-travis committed:
[UPD] Update base_multi_company.pot

pedrobaeza committed:
[FIX] base_multi_company: Fix tests

Odoo now checks if the user who creates the transient model is the same that is accessing
its data, so we need to create it with the same user.

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza committed:
[FIX] base_multi_company: Fix several things:

* Abstract model shouldn't have default for `company_ids`. It will be got by
  default `company_id` value.
* Record rule based on `company_id` makes the evaluation slower. Switch to `company_ids` field.
* Add `@api.depends` to `company_id` computation for proper refresh.
* Fix tests that were not correct but luckily previously suceeded. Now that other code has
  been removed, they have been uncovered.

oca-travis committed:
[UPD] Update base_multi_company.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza committed:
[MIG] partner_multi_company: Finish migration to 11.0

oca-travis committed:
[UPD] Update base_multi_company.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

florian-dacosta committed:
Migrate base_multi_company to v12

florian-dacosta committed:
Improve performance adding a techical field used in ir rules

florian-dacosta committed:
fixup! Migrate base_multi_company to v12

oca-travis committed:
[UPD] Update base_multi_company.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: multi-company-12.0/multi-company-12.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-12-0/multi-company-12-0-base_multi_company/

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 70.0% (7 of 10 strings)

Translation: multi-company-12.0/multi-company-12.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-12-0/multi-company-12-0-base_multi_company/zh_CN/

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (10 of 10 strings)

Translation: multi-company-12.0/multi-company-12.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-12-0/multi-company-12-0-base_multi_company/zh_CN/

max3903 committed:
[FIX] base_multi_company

OCA-git-bot committed:
[UPD] README.rst

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (10 of 10 strings)

Translation: multi-company-12.0/multi-company-12.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-12-0/multi-company-12-0-base_multi_company/hr/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (10 of 10 strings)

Translation: multi-company-12.0/multi-company-12.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-12-0/multi-company-12-0-base_multi_company/pt_BR/

alvarorib committed:
Translated using Weblate (Portuguese)

Currently translated at 90.0% (9 of 10 strings)

Translation: multi-company-12.0/multi-company-12.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-12-0/multi-company-12-0-base_multi_company/pt/

osi-scampbell committed:
[IMP] Turn Off IR Rule For Test

osi-scampbell committed:
[IMP] Flake8

JoanSForgeFlow committed:
[IMP] base_multi_company: black, isort

JoanSForgeFlow committed:
[MIG] base_multi_company: Migration to 13.0

oca-travis committed:
[UPD] Update base_multi_company.pot

OCA-git-bot committed:
[UPD] README.rst

pedrobaeza committed:
[IMP] base_multi_company: Pre-commit does this

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: multi-company-13.0/multi-company-13.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-13-0/multi-company-13-0-base_multi_company/

LoisRForgeFlow committed:
[13.0][FIX] base_multi_company: incompatibility error

if you have several companies active and any of them with
lower id than the current company, you can get multi-company
inconsistency errors.

E.g: when reserving stock pickings, product company is computed
and might be different than the move's company.

OCA-git-bot committed:
base_multi_company 13.0.1.0.1

LoisRForgeFlow committed:
[13.0][FIX] base_multi_company: consider force_company context key.

This is needed for transactions in which the company is forced, e.g
intercompany PO/SO synchronization.

OCA-git-bot committed:
base_multi_company 13.0.1.0.2

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (6 of 6 strings)

Translation: multi-company-13.0/multi-company-13.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-13-0/multi-company-13-0-base_multi_company/es_AR/

LoisRForgeFlow committed:
[13.0][FIX] base_multi_company: name_search

In some situations the 'in' operator is used with company_id in a
name_search. ORM does not convert to a proper WHERE clause when using
the 'in' operator.
e.g: ```
     WHERE "res_partner"."id" in (SELECT "res_partner_id"
     FROM "res_company_res_partner_rel" WHERE "res_company_id" IN (False, 1)
```
patching the args to expand the cumbersome args int a OR clause fix
the issue.
e.g: ```
     WHERE "res_partner"."id" not in (SELECT "res_partner_id"
             FROM "res_company_res_partner_rel"
             where "res_partner_id" is not null)
         OR  ("res_partner"."id" in (SELECT "res_partner_id"
             FROM "res_company_res_partner_rel" WHERE "res_company_id" IN 1)
```

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (6 of 6 strings)

Translation: multi-company-13.0/multi-company-13.0-base_multi_company
Translate-URL: https://translation.odoo-community.org/projects/multi-company-13-0/multi-company-13-0-base_multi_company/es/

rousseldenis committed:
[IMP] base_multi_company: black, isort, prettier

rousseldenis committed:
[14.0][MIG] base_multi_company

Use odoo-test-helper to test Abstract class

rousseldenis commented on base_multi_company/models/multi_company_abstract.py:
````diff
@@ -0,0 +1,78 @@
+        # ```
+        new_args = []
+        for arg in args:
+            if type(arg) == list and arg[:2] == ["company_id", "in"]:
````
@LoisRForgeFlow Could you give me the example of args and where it is generated in this case ?

I want to add further tests for this

LoisRForgeFlow replied:
I don't exactly recall now, but looking at the comment I added, I would say that something like `('company_id', 'in', [False, 1])` should be enough to reproduce the issue.

Let me know if that works, if not I would try to find the discussion when the problem was reported to me.

rousseldenis committed:
[14.0][IMP] Add test for name_search()

rousseldenis commented on base_multi_company/models/multi_company_abstract.py:
```diff
@@ -59,16 +59,17 @@ def _name_search(
-                new_args.extend(fix)
-            else:
-                new_args.append(arg)
+        if args:
```
args can be None.

LoisRForgeFlow replied:
that is more robust indeed

rousseldenis commented on base_multi_company/tests/test_multi_company_abstract.py:
```diff
@@ -100,6 +100,12 @@ def test_search_company_id(self):
         self.assertEqual(record, self.record_1)
 
+        name_result = self.test_model.name_search(
+            "test", [["company_id", "in", [self.company_2.id]]]
```
@LoisRForgeFlow args domain can be a list of a list and not a list of tuples ?

rousseldenis replied:
Ok, from js. But I think the problem was solved in 14 no ?

https://github.com/odoo/odoo/commit/69869ab681e2271ae802ca3ab7285ee83580208d#diff-fa4d9268d6e65e19aebec81c46038f0e496b91142588ed4d1c1bce7ff2338f2c

LoisRForgeFlow replied:
have you been able to reproduce in v13 with your test case, try to include a `False` in the "in" list?

rousseldenis commented:
WIP as I found an issue. Want to see if Odoo side or ours.

rousseldenis committed:
[14.0][FIX] base_multi_company: Use [] for domain on company_ids

rousseldenis committed:
[14.0][FIX] base_multi_company: Restore False for searches

Odoo seems to have a problem with subqueries, to avoid problems
with False values, store it in a Boolean field

rousseldenis committed:
[14.0][IMP] base_multi_company: Add hook for new field

rousseldenis committed:
[14.0][FIX] base_multi_company: Add requirements.txt

rousseldenis committed:
[14.0][FIX] base_multi_company: Compute no_company_ids as sudo

rousseldenis commented:
This one is ready and working with product_multi_company.

As there is a problem when evaluating domain like `['|', ('company_ids', '=', False), ('company_ids', 'in', company_ids)]`.  The first part is added as subquery and for an unknown reason does not return correct records.

Something has changed for v14 there : https://github.com/odoo/odoo/commit/69869ab681e2271ae802ca3ab7285ee83580208d#diff-fa4d9268d6e65e19aebec81c46038f0e496b91142588ed4d1c1bce7ff2338f2c

The idea here is to get a field that means 'no company is selected on record' as a boolean and evaluate it like that.

rousseldenis committed:
[14.0][IMP] base_multi_company: Avoid warnings in logs

pedrobaeza committed:
[FIX] base_multi_company: Make company_id writeable again

Not being writeable, we have warnings on constraints over this field
(it's used for example in `sale` module for `product.template` model).

Giving an inverse to the field, we avoid this problem and give
compatibility to any other module that writes this field.

pedrobaeza commented:
Can you please take into account #313 ?

pedrobaeza commented:
I see that your last commit attempts to fix the same thing, but your patch seems incomplete for preserving compatibility.

rousseldenis commented:
@OCA/core-maintainers

rousseldenis commented:
> Can you please take into account #313 ?

Why company_id should be writeable, value should not to be taken into account.

rousseldenis commented:
> I see that your last commit attempts to fix the same thing, but your patch seems incomplete for preserving compatibility.

Do you have written a test case that shows company_ids could be void ? As your create and write part seem not useful https://github.com/OCA/multi-company/pull/313/files#diff-768c66d1106f5fdcb827fc104212e9b76a2739fae84f770ffe2687ec3bea1c4bR52

pedrobaeza commented:
Yes, it should be for compatibility with other modules that are not aware of this new field `company_ids`. Starting with, if you don't do it writeable, you will have a warning when installed with `sale`, as the field is used for a constraint. We traced other problems to this fact, and that's why we added the PR, and since then, everything goes smooth. If any module writes in `company_id`, it's reflected in company_ids as if you have put only one company (or none) there.

rousseldenis commented:
> Yes, it should be for compatibility with other modules that are not aware of this new field `company_ids`. Starting with, if you don't do it writeable, you will have a warning when installed with `sale`, as the field is used for a constraint. We traced other problems to this fact, and that's why we added the PR, and since then, everything goes smooth. If any module writes in `company_id`, it's reflected in company_ids as if you have put only one company (or none) there.

Could you give the real case as we didn't have problems ?

pedrobaeza commented:
I don't have the trace right now, but arose on integration tests when combined with several modules. I have already put you an example with `sale` module (it's just a warning, but it also means that you are disabling that constraint not doing it writeable). The create/write overriding is for giving always priority to `company_ids` definition instead of `company_id` if both are specified, which will mean any module aware of the multi company thing (or the module itself) can operate transparently to this `company_id` writable thing.

rousseldenis commented:
> I don't have the trace right now, but arose on integration tests when combined with several modules. I have already put you an example with `sale` module (it's just a warning, but it also means that you are disabling that constraint not doing it writeable). The create/write overriding is for giving always priority to `company_ids` definition instead of `company_id` if both are specified, which will mean any module aware of the multi company thing (or the module itself) can operate transparently to this `company_id` writable thing.

Ok, yes, I begin to see what  you want to say.

I will add your code and add tests too in another commit if you wnat to backport it

pedrobaeza commented:
The existing tests of the module lead me to add the create/write overriding analyzing the fail, so they are useful also for this, but it's OK to add some more if you see uncovered things.

rousseldenis committed:
[14.0][IMP] base_multi_company: Allow non-aware multi-company behaviours

rousseldenis committed:
[14.0][IMP] base_multi_company: Add depends_context for env.company

GSLabIt reviewed (approved):
LGTM

pedrobaeza commented:
I'm missing the create/write part for discarding the double write of both `company_id` and `company_ids`.

pedrobaeza commented:
Sorry, wrong button

elvise commented:
> I'm missing the create/write part for discarding the double write of both `company_id` and `company_ids`.

@pedrobaeza how can we proceed towards the merge?

pedrobaeza commented:
@rousseldenis can you include such changes?

francesco-ooops commented:
hi @rousseldenis , any plans to finish this one?

rousseldenis commented:
> hi @rousseldenis , any plans to finish this one?

I'll try but I need to dig some time

LoisRForgeFlow commented:
@rousseldenis A friendly reminder here, have you got the time to finish this?

GSLabIt commented:
@rousseldenis did u have the chance to dive into this?

rousseldenis commented:
@GSLabIt @LoisRForgeFlow  @pedrobaeza This is done

pedrobaeza commented on base_multi_company/models/multi_company_abstract.py:
```diff
@@ -0,0 +1,126 @@
+
+        for record in self:
+            company = record.company_id
+            if company and company not in record.company_ids:
```
Wait, this is not like my PR. We can't do make conditional that way, or we are modifying what is intended to do assigning a company, that is to change all the allowed companies to this one.

rousseldenis replied:
That was done with that purpose. See comment before.

If people has no access to other companies, he cannot remove them.

pedrobaeza replied:
But this is not OK. The intention to add the writeable company_id is to do exactly what is expected to do: limit the record to such company. This is working perfectly in v13. If not, please specify steps to reproduce a problem with this behavior.

rousseldenis replied:
The behaviour changed in 13 from which it was at origin:

https://github.com/OCA/multi-company/commit/f5f3270ee8ab8139bd7344d1087f8872537332c7

You reintroduced the inverse BUT changing the behaviour.

See the original commit at creation in 10:

https://github.com/OCA/multi-company/commit/7051f69ec672aea10bbe3518916927fdbc15aba9#diff-768c66d1106f5fdcb827fc104212e9b76a2739fae84f770ffe2687ec3bea1c4bR45

rousseldenis replied:
My change was, at purpose, restablishing the original behaviour.

pedrobaeza replied:
Such original behavior was wrong, as it's modifying whole standard behavior. If any standard module not being aware of this multi-company think wants to set `company_id`, the expected behavior should be that the only allowed company must be that one, not that one + others not expected.

You are not adding any step to reproduce any problem with the behavior I'm mentioning. I will be glad to change my mind if you find steps that provokes a bug using my approach.

On contrary, this change will make lines like this one:

https://github.com/odoo/odoo/blob/4ad64719316bcd59c3961ed3b19d560a96215ff7/odoo/addons/base/models/res_users.py#L478

to not behave as expected.

rousseldenis replied:
> You are not adding any step to reproduce any problem with the behavior I'm mentioning. I will be glad to change my mind if you find steps that provokes a bug using my approach.

On the contrary, I want a test from you that show me the wrong behaviour you 'corrected' as you didn't provided a test...

rousseldenis replied:
@JoanSForgeFlow @LoisRForgeFlow @JordiBForgeFlow @JasminSForgeFlow

As you use/migrated this one in several versions, could you help here ? What is the expected behaviour with the write on company_id field ?

rousseldenis replied:
I'll fix this next week. Need to sleep on it :-)

pedrobaeza replied:
Thanks for considering it.

rousseldenis replied:
@pedrobaeza I've taken your changes entirely.

I need now to improve README

pedrobaeza replied:
Great, thanks. Let me know when this can be merged.

rousseldenis committed:
[14.0][IMP] base_multi_company: Avoid one recordset update + add tests

rafaelbn reviewed (approved):
👍 Thank you!

rousseldenis committed:
[14.0][IMP] base_multi_company: Improve module README

rousseldenis commented:
@pedrobaeza @Cedric-Pigeon @LoisRForgeFlow This one is ready

pedrobaeza commented on base_multi_company/README.rst:
```diff
@@ -0,0 +1,108 @@
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module will provide a way to change the way Odoo manages a 'multi-company'
+implentation.
```
```suggestion
implementation.
```

pedrobaeza commented on base_multi_company/README.rst:
```diff
@@ -0,0 +1,108 @@
+  - Limit access to that company if set.
+  - not limiting access to any company if not set.
+
+This module breaks that in order to introduce a fine company access.
```
break sounds very hard

```suggestion
This module changes that in order to introduce a finer company access.
```

pedrobaeza commented on base_multi_company/README.rst:
```diff
@@ -0,0 +1,108 @@
+This module is not doing anything by its own but provide a transversal implementation
+for further ones.
+e.g.: If you want to implement OCA multi-company behaviour for products, install
+also the 'product_multi_company' module.
```
You can also mention `partner_multi_company` as the 2 well known modules implementing this.

pedrobaeza commented:
Thanks, and I'm going to launch the bot with `patch` option for satisfying you although I don't like that glitch :)

/ocabot merge patch

rousseldenis commented:
> Thanks, and I'm going to launch the bot with `patch` option for satisfying you although I don't like that glitch :)
> 
> /ocabot merge patch

contribution is a location of compromises :smile: :rocket:
